### DISCUSSION TITLE
Crazy radio question

### DISCUSSION THREAD

**Discussion by kalib-loop** [bitcraze/discussions#394] [Q&A]
Hello, I have a question regarding the Crazy radio specifically in the Loco tracking bundle. in the Loco tracking bundle why is it there are 3 radios rather than 1? I would assume this is due to the radio only have a certain amount of bandwidth and can only take impute from so many Crazyflies at one time. If this is true then does this scale linearly and how many crazyflies can each radio handle?

**Comment by ataffanel** [comment]
This is correct, the inclusion of more than one radio in the swarm bundles is because the bandwidth is a limited resource that needs to be shared between Crazyflies.

How many Crazyflie per radio depends very much of the use-case: if only sending high-level setpoint to the Crazyflies (ie. having the crazyflie mostly autonomous) and the log are kept at a minimum, one radio for 10 Crazyflies can work well. In more common use-case though, with reasonable logging and control rate, we tend to limit to 3-4 Crazyflie per Crazyradio (hence the 3 Crazyradios). It is of course possible to load the link in such a way that one radio per Crazyflie is required, but this would most likely be for logging a lot of data at high rate to debug a specific problem, not when flying as a swarm.

As for the scaling, there is a drop of performance when using a radio from 1 to many Crazyflie. However radios are mostly independent so splitting 4 Crazyflies on two radios rather than on one will give twice the bandwidth.

**Reply by kalib-loop** [reply]
Alright that makes a lot of sense. 

Now in the example of the Loco tracking system, are the anchors also sending information to the crazy radio or is it the crazyflie itself sending tracking information? Is this dependent on which protocol your using for the UWB? I would assume in TWR the anchors are also part of the conversation with the radio but in TDOA maybe its only the crazyflies communicating with the radios.

Thank you for all the help so far it has been quite useful :)

**Reply by kalib-loop** [reply]
Also does reducing the bitrate in long range mode allow the crazy radio to handle more Crazyflies?

**Reply by ataffanel** [reply]
The Loco tracking is using Ultra Wide Band (UWB) radio which is completely independent from the Crazyradio, Crazyradio cannot talk to UWB and vice-versa.

In a typical Loco positioning system, the position is acquired and used in the Crazyflie and it is optionally sent to the ground using the logging subsystem via the Crazyradio.

Reducing the Crazyradio bit-rate will make the radio able to handle less Crazyflies, not more, since the total bandwidth will be reduced. When using 2Mbps the rule of thumb is that one radio can handle about 800 packets per seconds shared among the Crazyflies. This will be reduced a bit at 1Mbps.

**Reply by kalib-loop** [reply]
Got it. 
what is the refresh rate of the Loco positioning system? 
If each crazyflies is updating their position 50 times/second and sending that back to ground over the logging system, in simplest terms is 1 update 1 packet sent?

**Reply by ataffanel** [reply]
The packet rate (in simple terms we get one new measurement per UWB packet) of the Loco positioning in TDoA3 mode is of about 200Pk/s, However the ranging data are pushed in a kalman filter that has an update rate of 100Hz so the Crazyflie generates a new pose estimate at 100Hz.

The logging rate is independent of the position update rate (they are asynchronous) and so should rather be set depending of your use-case. Logging at 50Hz will indeed generates 50 packets per seconds.
